### PR TITLE
docs: Document page tables, visibility, per-CPU guards

### PR DESCRIPTION
**Pull request by Zildj1an**
Document memory management of page tables, page visibility and per-CPU mapping guards. This is related to issue https://github.com/coconut-svsm/svsm/issues/74 Document COCONUT-SVSM and continues the mm documentation efforts that started with the merged https://github.com/coconut-svsm/svsm/pull/243. For the purpose of that issue, I will consider the memory subsystem sufficiently documented after this PR (even if not every single detail has been documented) and would like to shift focus to documenting `protocols`.

### CODE REVIEW

File: kernel/src/mm/pagetable.rs [hunk 1]
```diff
@@ -294,12 +354,21 @@ impl PageTable {
...
+    ///
+    /// # Returns
+    /// A `Mapping` representing the found mapping.
     fn walk_addr_lvl0(page: &mut PTPage, vaddr: VirtAddr) -> Mapping<'_> {
         let idx = PageTable::index::<0>(vaddr);
 
         Mapping::Level0(&mut page[idx])
     }
 
+    /// Walks a page table at level 1 to find a mapping.
```
Comment (stefano-garzarella) [inline kernel/src/mm/pagetable.rs:371]: We are documenting parameters and return values for `walk_addr_lvl0` and `walk_addr_lvl3`, so maybe we can do also for `walk_addr_lvl1` and `walk_addr_lvl2`

File: kernel/src/mm/pagetable.rs [hunk 2]
```diff
@@ -322,6 +392,14 @@ impl PageTable {
         };
     }
 
+    /// Walks the page table to find a mapping for a given virtual address.
```
Comment (stefano-garzarella) [inline kernel/src/mm/pagetable.rs:395]: Should we mention "level 3"?
I mean something like `Walks the page table at level 3 to find ...`

File: kernel/src/mm/pagetable.rs [hunk 3]
```diff
@@ -692,6 +896,8 @@ impl PageTable {
         }
     }
 
+    /// Populates this paghe table with the contents of the given.
```
Comment (stefano-garzarella) [inline kernel/src/mm/pagetable.rs:899]:
```suggestion
    /// Populates this paghe table with the contents of the given
```

File: kernel/src/mm/pagetable.rs [hunk 4]
```diff
@@ -793,6 +1039,7 @@ impl RawPageTablePart {
         Some(unsafe { &mut *address.as_mut_ptr::<PTPage>() })
     }
 
+    /// Frees a level 1 page table.
```
Comment (stefano-garzarella) [inline kernel/src/mm/pagetable.rs:1042]:
```suggestion
    /// Frees a level 1 page table.
    ///
    /// # Parameters
    /// - `page`: A reference to the level 1 page table to be freed.
```

File: kernel/src/mm/pagetable.rs [hunk 5]
```diff
@@ -920,6 +1232,16 @@ impl RawPageTablePart {
...
+    /// Unmaps a 2MB page.
+    ///
+    /// # Parameters
+    /// - `vaddr`: The virtual address of the mapping to unmap.
+    ///
+    /// # Returns
+    /// An optional [`PTEntry`] representing the unmapped page table entry.
+    /// # Panics
+    ///
+    /// Panics if `vaddr` is not memory aligned.
```
Comment (stefano-garzarella) [inline kernel/src/mm/pagetable.rs:1244]:
```suggestion
    /// An optional [`PTEntry`] representing the unmapped page table entry.
    ///
    /// # Panics
    /// Panics if `vaddr` is not memory aligned.
```

File: kernel/src/mm/ptguards.rs [hunk 6]
```diff
@@ -24,6 +25,24 @@ pub struct PerCPUPageMappingGuard {
...
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing the [`PerCPUPageMappingGuard`] if successful,
+    /// or an `SvsmError` if an error occurs.
+    ///
+    /// # Panics
+    ///
+    /// Panics if either `paddr_start`, the size, or `paddr_end`, are not
+    /// aligned
```
Comment (stefano-garzarella) [inline kernel/src/mm/ptguards.rs:45]:
```suggestion
    /// aligned.
```

**Review by stefano-garzarella — COMMENTED** [review]
Thanks for this work!

I left a few comments on a few things that jumped out at me, feel free to ignore them if they are too nitpicky.

**Comment by Zildj1an** [comment]
I have applied your feedback @00xc @stefano-garzarella

**Review by 00xc — APPROVED** [review]

**Pull request merged**